### DISCUSSION TITLE
What happens if the implementation of IHostedService.StopAsync method does not respond to cancellation requests ?

### DISCUSSION THREAD

**Discussion by EnricoMassone** [dotnet/aspnetcore#23742] [General] [answered]
Hi,

we are implementing a .NET core 3.1 worker service and we need to register several hosted service to implement the required business. 

[This documentation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/host/hosted-services?view=aspnetcore-3.1&tabs=visual-studio#timed-background-tasks) explains that the [StopAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice.stopasync?view=dotnet-plat-ext-3.1#Microsoft_Extensions_Hosting_IHostedService_StopAsync_System_Threading_CancellationToken_) method of the [IHostedService](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice?view=dotnet-plat-ext-3.1) interface should be able to listen to cancellation requests issued after a default 5 seconds timeout. 

Based on my understanding, the [StopAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice.stopasync?view=dotnet-plat-ext-3.1#Microsoft_Extensions_Hosting_IHostedService_StopAsync_System_Threading_CancellationToken_) method is called by the host during the shutdown process of the host itself. 

In one of our hosted services we need to call an async method during the hosted service stop process; unfortunately this method does **not** support cancellation, so there is not a simple way for us to implement the [StopAsync](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.ihostedservice.stopasync?view=dotnet-plat-ext-3.1#Microsoft_Extensions_Hosting_IHostedService_StopAsync_System_Threading_CancellationToken_) method in the proper way. 

This is our scenario: 

```
public sealed class MyHostedService : IHostedService 
{
     public async Task StopAsync(CancellationToken token)
     {
          _logger.LogInformation("Stopping the service....");

           await DoSomethingAsync(); // this method does not support cancellation, so I can't pass the cancellation token here

          _logger.LogInformation("Service was stopped");
     }
}
```

What's going to happen at the host level in this scenario ? 

Is this implementation "dangerous" in any sort ? 

Is this going to block the host shutdown process in case of a long time needed to complete the execution of `StopAsync` ? 

The simplest solution to this issue is implementing the class `MyHostedService`  by deriving from the [BackgroundService](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.backgroundservice?view=dotnet-plat-ext-3.1)  base class, but I would like to fully understand the possible implications of a long running `StopAsync` implementation at the host level. 

Thanks for helping.

**Comment by Tratcher** [comment, marked as answer]
Yes, this could prevent the host from shutting down when requested. I'd rethink how much work you do during shutdown if it's likely to take more than 5 seconds.

**Reply by EnricoMassone** [reply]
@Tratcher thanks for your answer. So, basically the idea is that an hosted service implementation should be able to stop quickly, right ?

**Reply by Tratcher** [reply]
Yes, gracefully or abortively.
